File: src/speech/chat-input.ts
import type { BackendEventEnvelope, ProtocolChatEvent } from '../minecraft/contracts.js'
import type { ChatInputContext, PlayerChatMessage } from './contracts.js'

export function interpretPlayerChat(
  event: BackendEventEnvelope<ProtocolChatEvent>,
  context: ChatInputContext,
): PlayerChatMessage | undefined {
  if (event.kind !== 'chat' || event.payload.position !== 'chat' || !event.payload.senderUsername) return undefined
  const sender = event.payload.senderUsername
  const explicitName = mentionsName(event.payload.plainText, context.companionUsername)
  const ongoing = equalName(context.conversationActiveWith, sender)
  const onlinePlayers = context.onlinePlayerUsernames.filter(name => !equalName(name, context.companionUsername))
  // 单独在场规则是一个临时的中间层决定（按 W09 在此记录理由）：两个人的场合说话不必点名，
  // 对象不言自明。规则属于「唯一在场」这一处境，不属于任何特定玩家——同条件同规则（A10）。
  // 预期归宿：帧携带在线人数与消息统计，关注与关系由 AI 写在记忆和关注列表里、由 AI 自己
  // 决定听谁（W08a）；届时本规则应被到达/关注机制取代或重新评估，而不是长成第二套特权。
  const singleParty = onlinePlayers.length === 1 && equalName(onlinePlayers[0], sender)
  const evidence = [
    ...(explicitName ? ['explicit_name' as const] : []),
    ...(ongoing ? ['ongoing_conversation' as const] : []),
    ...(singleParty ? ['single_party' as const] : []),
  ]
  const addressed = evidence.length > 0
  return {
    protocol: 'mineintent.player-chat.v1',
    sourceEventId: event.id,
    occurredAt: event.occurredAt,
    sender: { username: sender },
    text: event.payload.plainText,
    ...(event.payload.verified === undefined ? {} : { verified: event.payload.verified }),
    addressing: { addressedToCompanion: addressed, evidence: evidence.length ? evidence : ['not_addressed'] },
    world: { worldId: event.worldId, ...(event.dimension ? { dimension: event.dimension } : {}), connectionEpoch: event.connectionEpoch },
  }
}

function equalName(a: string | undefined, b: string): boolean { return a?.toLocaleLowerCase() === b.toLocaleLowerCase() }
function mentionsName(text: string, name: string): boolean {
  return new RegExp(`(?:^|[@＠\\s，,：:])${escapeRegex(name)}(?:$|[\\s，,：:！!.?？])`, 'iu').test(text)
}
function escapeRegex(value: string): string { return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') }

File: crates/middle/src/speech/chat_input.rs
use mineintent_contracts::minecraft::{
    BackendEventEnvelope, BackendEventKind, ChatPosition, ProtocolChatEvent,
};

use super::{
    Addressing, AddressingEvidence, ChatInputContext, PlayerChatMessage, PlayerChatProtocol,
    PlayerChatSender, PlayerChatWorld,
};

pub fn interpret_player_chat(
    event: &BackendEventEnvelope<ProtocolChatEvent>,
    context: &ChatInputContext,
) -> Option<PlayerChatMessage> {
    if event.kind != BackendEventKind::Chat || event.payload.position != Some(ChatPosition::Chat) {
        return None;
    }
    let sender = event
        .payload
        .sender_username
        .as_deref()
        .filter(|sender| !sender.is_empty())?;

    let explicit_name = mentions_name(&event.payload.plain_text, &context.participant_username);
    let ongoing = context
        .conversation_active_with
        .as_deref()
        .is_some_and(|name| equal_name(name, sender));
    let mut other_players = context
        .online_player_usernames
        .iter()
        .filter(|name| !equal_name(name, &context.participant_username));
    let only_other = other_players.next();
    let single_party =
        only_other.is_some_and(|name| equal_name(name, sender)) && other_players.next().is_none();

    let mut evidence = Vec::with_capacity(3);
    if explicit_name {
        evidence.push(AddressingEvidence::ExplicitName);
    }
    if ongoing {
        evidence.push(AddressingEvidence::OngoingConversation);
    }
    if single_party {
        evidence.push(AddressingEvidence::SingleParty);
    }
    let addressed_to_participant = !evidence.is_empty();
    if evidence.is_empty() {
        evidence.push(AddressingEvidence::NotAddressed);
    }

    Some(PlayerChatMessage {
        protocol: PlayerChatProtocol::V1,
        source_event_id: event.id.clone(),
        occurred_at: event.occurred_at.clone(),
        sender: PlayerChatSender {
            username: sender.to_owned(),
        },
        text: event.payload.plain_text.clone(),
        verified: event.payload.verified,
        addressing: Addressing {
            addressed_to_participant,
            evidence,
        },
        world: PlayerChatWorld {
            world_id: event.world_id.clone(),
            dimension: event.dimension.clone(),
            connection_epoch: event.connection_epoch,
        },
    })
}

fn equal_name(left: &str, right: &str) -> bool {
    left.to_lowercase() == right.to_lowercase()
}

fn mentions_name(text: &str, name: &str) -> bool {
    let text: Vec<char> = text.to_lowercase().chars().collect();
    let name: Vec<char> = name.to_lowercase().chars().collect();
    if name.is_empty() || name.len() > text.len() {
        return false;
    }

    (0..=text.len() - name.len()).any(|start| {
        text[start..start + name.len()] == name
            && (start == 0 || is_name_prefix(text[start - 1]))
            && (start + name.len() == text.len() || is_name_suffix(text[start + name.len()]))
    })
}

fn is_name_prefix(character: char) -> bool {
    is_javascript_whitespace(character) || matches!(character, '@' | '＠' | '，' | ',' | '：' | ':')
}

fn is_name_suffix(character: char) -> bool {
    is_javascript_whitespace(character)
        || matches!(
            character,
            '，' | ',' | '：' | ':' | '！' | '!' | '.' | '?' | '？'
        )
}

fn is_javascript_whitespace(character: char) -> bool {
    matches!(
        character,
        '\u{0009}'..='\u{000D}'
            | '\u{0020}'
            | '\u{00A0}'
            | '\u{1680}'
            | '\u{2000}'..='\u{200A}'
            | '\u{2028}'
            | '\u{2029}'
            | '\u{202F}'
            | '\u{205F}'
            | '\u{3000}'
            | '\u{FEFF}'
    )
}
